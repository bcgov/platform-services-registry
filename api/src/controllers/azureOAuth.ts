import axios from "axios";
import { logger } from "@bcgov/common-nodejs-utils";
// import { Response } from "express";
import qs from "qs";
import { Request, Response } from "express";

export const fetchAzureAccessToken = async (req: Request, res: Response): Promise<void> => {
  try {
    const options = {
      Host: "login.microsoftonline.com",
      client_id: `${process.env.AZURE_CLIENT_ID}`,
      scope: "https://graph.microsoft.com/.default",
      client_secret: process.env.AZURE_CLIENT_SECRET,
      grant_type: "client_credentials",
    };
    const url: string = `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}/oauth2/v2.0/token`;
    axios
      .post(url, qs.stringify(options))
      .then((response) => {
        if (!response.data.access_token) {
          throw Error(
            "a result was returned from Microsoft's token endpoint, but it doesn't seem to have an access token included"
          );
        }
        res.status(200).json(response.data.access_token);
      })
      .catch((error) => {
        logger.info(`error: ${error}`);
        res.status(500);
      });
  } catch (err) {
    logger.error(`err = ${err.message}`);
  }
};

export default fetchAzureAccessToken;
