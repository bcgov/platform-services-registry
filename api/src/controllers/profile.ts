//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import shared from '../libs/shared';
import { isNotAuthorized, validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserId(user.id);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    res.status(200).json(record);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix. If it exists, this
  // overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();

    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    categoryId,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
  } = body;

  try {
    const record = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      categoryId,
      busOrgId,
      prioritySystem,
      userId: record.userId,
      namespacePrefix: record.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
    };

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(profileId);

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

    res.status(202).end();
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ContactModel, ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const results = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
