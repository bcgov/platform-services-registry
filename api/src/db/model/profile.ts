// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-28.
//

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export interface ProjectProfile extends CommonFields {
  name: string,
  description: string,
  busOrgId: number,
  userId: number,
  namespacePrefix: string,
  prioritySystem?: boolean,
  criticalSystem?: boolean,
  notificationEmail?: boolean,
  notificationSms?: boolean,
  notificationMsTeams?: boolean,
  paymentBambora?: boolean,
  paymentPayBc?: boolean,
  fileTransfer?: boolean,
  fileStorage?: boolean,
  geoMappingWeb?: boolean,
  geoMappingLocation?: boolean,
  schedulingCalendar?: boolean,
  schedulingAppointments?: boolean,
  idmSiteMinder?: boolean,
  idmKeycloak?: boolean,
  idmActiveDir?: boolean,
  other: string,
}

export default class ProfileModel extends Model {
  table: string = 'profile';
  requiredFields: string[] = [
    'name',
    'description',
    'busOrgId',
    'prioritySystem',
    'userId',
    'namespacePrefix',
  ];
  pool: Pool;

  constructor(pool: any) {
    super();
    this.pool = pool;
  }

  async create(data: ProjectProfile): Promise<ProjectProfile> {
    const query = {
      text: `
        INSERT INTO ${this.table}
          (name, description, bus_org_id, priority_system,
            critical_system, user_id, namespace_prefix,
            notification_email, notification_sms, notification_ms_teams,
            payment_bambora, payment_pay_bc, file_transfer, file_storage,
            geo_mapping_web, geo_mapping_location, scheduling_calendar,
            scheduling_appointments, idm_site_minder,
            idm_keycloak, idm_active_dir,
            other)
          VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12,
            $13, $14, $15, $16, $17, $18, $19, $20, $21, $22) RETURNING *;`,
      values: [
        data.name.trim(),
        data.description,
        data.busOrgId,
        data.prioritySystem ? data.prioritySystem : false,
        data.criticalSystem ? data.criticalSystem : false,
        data.userId,
        data.namespacePrefix,
        data.notificationEmail ? data.notificationEmail : false,
        data.notificationSms ? data.notificationSms : false,
        data.notificationMsTeams ? data.notificationMsTeams : false,
        data.paymentBambora ? data.paymentBambora : false,
        data.paymentPayBc ? data.paymentPayBc : false,
        data.fileTransfer ? data.fileTransfer : false,
        data.fileStorage ? data.fileStorage : false,
        data.geoMappingWeb ? data.geoMappingWeb : false,
        data.geoMappingLocation ? data.geoMappingLocation : false,
        data.schedulingCalendar ? data.schedulingCalendar : false,
        data.schedulingAppointments ? data.schedulingAppointments : false,
        data.idmSiteMinder ? data.idmSiteMinder : false,
        data.idmKeycloak ? data.idmKeycloak : false,
        data.idmActiveDir ? data.idmActiveDir : false,
        data.other,
      ],
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create project profile`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findByPrefix(prefix: string): Promise<ProjectProfile> {
    const query = {
      text: `
        SELECT * FROM ${this.table}
          WHERE namespace_prefix = $1;
      `,
      values: [
        prefix,
      ],
    };

    try {
      const results = await this.runQuery(query);

      return results.pop();
    } catch (err) {
      const message = `Unable to find Profile by prefix ${prefix}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async update(profileId, data: ProjectProfile): Promise<ProjectProfile> {
    const values: any[] = [];
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            name = $1, description = $2, bus_org_id = $3,
            priority_system = $4, notification_email = $5, notification_sms = $6,
            notification_ms_teams = $7, payment_bambora = $8, payment_pay_bc = $9, 
            file_transfer = $10, file_storage = $11, geo_mapping_web = $12,
            geo_mapping_location = $13, scheduling_calendar = $14, scheduling_appointments = $15,
            idm_site_minder = $16, idm_keycloak = $17,
            idm_active_dir = $18, other = $19
          WHERE id = ${profileId}
          RETURNING *;`,
      values,
    };

    try {
      const record = await this.findById(profileId);
      const aData = { ...record, ...data };
      query.values = [
        aData.name,
        aData.description,
        aData.busOrgId,
        aData.prioritySystem ? aData.prioritySystem : false,
        aData.notificationEmail ? aData.notificationEmail : false,
        aData.notificationSms ? aData.notificationSms : false,
        aData.notificationMsTeams ? aData.notificationMsTeams : false,
        aData.paymentBambora ? aData.paymentBambora : false,
        aData.paymentPayBc ? aData.paymentPayBc : false,
        aData.fileTransfer ? aData.fileTransfer : false,
        aData.fileStorage ? aData.fileStorage : false,
        aData.geoMappingWeb ? aData.geoMappingWeb : false,
        aData.geoMappingLocation ? aData.geoMappingLocation : false,
        aData.schedulingCalendar ? aData.schedulingCalendar : false,
        aData.schedulingAppointments ? aData.schedulingAppointments : false,
        aData.idmSiteMinder ? aData.idmSiteMinder : false,
        aData.idmKeycloak ? aData.idmKeycloak : false,
        aData.idmActiveDir ? aData.idmActiveDir : false,
        aData.other,
      ];
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create project profile`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  };

  async delete(profileId): Promise<ProjectProfile> {
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            archived = true
          WHERE id = ${profileId}
          RETURNING *;
      `,
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to archive project profile`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  };

  async addContactToProfile(profileId: number, contactId: number): Promise<void> {
    const values: any[] = [];
    const table = 'profile_contact';
    const query = {
      text: `
        INSERT INTO ${table}
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;`,
      values,
    };

    try {
      query.values = [
        profileId,
        contactId,
      ];

      const results = await this.runQuery(query);

      return results.pop();
    } catch (err) {
      const message = `Unable to link contact ${contactId} to profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async isNamespacePrefixUnique(prefix: string): Promise<boolean> {
    const query = {
      text: `
        SELECT COUNT(*) FROM ${this.table}
          WHERE namespace_prefix = $1;`,
      values: [
        prefix,
      ],
    };

    try {
      const results = await this.runQuery(query);

      return Number(results.pop().count) === 0 ? true : false;
    } catch (err) {
      const message = `Unable to lookup namespace prefix ${prefix}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findProfilesByUserId(id: number): Promise<any> {
    const query = {
      text: `
        SELECT * FROM ${this.table}
          WHERE user_id = $1 AND archived = false;`,
      values: [id],
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to fetch Profile(s) with User Id ${id}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }
}
