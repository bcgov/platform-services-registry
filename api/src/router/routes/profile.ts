//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-21.
//

'use strict';

import { asyncMiddleware } from '@bcgov/common-nodejs-utils';
import express from 'express';
import { requestContactEdit } from '../../controllers/contact';
import { archiveProfileNamespace, createNamespace, fetchProfileNamespace, fetchProfileNamespaces, fetchProfileQuotaOptions, requestProfileQuotaEdit, updateProfileNamespace } from '../../controllers/namespace';
import { addContactToProfile, archiveProjectProfile, createProjectProfile, fetchAllProjectProfiles, fetchProfileContacts, fetchProfileEditRequests, fetchProjectProfile, updateProjectProfile } from '../../controllers/profile';

const router = express.Router();

// Profiles
router.post('/', asyncMiddleware(createProjectProfile));
router.get('/', asyncMiddleware(fetchAllProjectProfiles));
router.get('/:profileId', asyncMiddleware(fetchProjectProfile));
router.put('/:profileId', asyncMiddleware(updateProjectProfile));
router.delete('/:profileId', asyncMiddleware(archiveProjectProfile));

// Namespace
router.post('/:profileId/namespace', asyncMiddleware(createNamespace));
router.get('/:profileId/namespace', asyncMiddleware(fetchProfileNamespaces));
router.get('/:profileId/namespace/:namespaceId', asyncMiddleware(fetchProfileNamespace));
router.put('/:profileId/namespace/:namespaceId', asyncMiddleware(updateProfileNamespace));
router.delete('/:profileId/namespace/:namespaceId', asyncMiddleware(archiveProfileNamespace));

// Contacts
router.get('/:profileId/contacts', asyncMiddleware(fetchProfileContacts));
router.post('/:profileId/contact/:contactId', asyncMiddleware(addContactToProfile));
router.post('/:profileId/contact-edit', asyncMiddleware(requestContactEdit));

// Quota-edit
router.get('/:profileId/quota-edit', asyncMiddleware(fetchProfileQuotaOptions));
router.post('/:profileId/quota-edit', asyncMiddleware(requestProfileQuotaEdit));
router.get('/:profileId/request', asyncMiddleware(fetchProfileEditRequests));
export default router;
