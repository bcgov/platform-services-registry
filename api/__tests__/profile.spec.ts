//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { fetchProfileContacts, requestProfileContactsEdit, requestProjectProfileEdit } from '../src/controllers/profile';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-profiles.json');
const selectProfiles = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/user-template.json');
const userRequest = JSON.parse(fs.readFileSync(p1, 'utf8'));

const p2 = path.join(__dirname, 'fixtures/select-profile-contacts.json');
const selectProfilesContacts = JSON.parse(fs.readFileSync(p2, 'utf8'));

const client = new Pool().connect();

jest.mock('../src/libs/fulfillment', () => {
  const p3 = path.join(__dirname, 'fixtures/provisioning-context.json');
  const natsContext = JSON.parse(fs.readFileSync(p3, 'utf8'));
  const natsSubject = 'edit';
  return {
    fulfillEditRequest: jest.fn().mockResolvedValue({ natsContext, natsSubject }),
  };
});

describe('Profile event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();

    ex = new FauxExpress();
  });

  it('A profiles contacts are returned', async () => {
    const req = {
      params: { profileId: 1 },
      user: userRequest,
    };

    client.query.mockReturnValueOnce({ rows: [selectProfiles[0]] });
    client.query.mockReturnValueOnce({ rows: selectProfilesContacts });

    // @ts-ignore
    await fetchProfileContacts(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch single profile contacts should throw', async () => {
    const req = {
      params: { profileId: 1 },
      user: userRequest,
    };
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(fetchProfileContacts(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('A project-profile edit request is created', async () => {
    const req = {
      params: { profileId: 118 },
      body: selectProfiles[2],
    }

    client.query.mockResolvedValue({ rows: [{ count: '0' }] });

    await requestProjectProfileEdit(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.end).toBeCalled();
  });


  it('A contact edit request is created', async () => {
    const body = {
      productOwner: {
        roleId: 1,
        id: 233,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@example.com',
        githubId: 'jane1100',
      },
      technicalContact: {
        roleId: 2,
        id: 234,
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        githubId: 'john1100',
      },
    };
    const req = {
      params: { profileId: 118 },
      body,
    }

    client.query.mockResolvedValue({
      roleId: 1,
      id: 233,
      firstName: 'Jane',
      lastName: 'Doe',
      email: 'jane.doe@example.com',
      githubId: 'jane1100',
    });

    client.query.mockResolvedValue({
      productOwner: {
        userId: 'jane1100',
        provider: 'github',
        email: 'jane.doe@example.ca',
      },
      technicalContact: {
        userId: 'john1100',
        provider: 'github',
        email: 'john.doe@example.com',
      },
    });

    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });
    client.query.mockReturnValueOnce({ rows: [{ count: '0' }] });

    await requestProfileContactsEdit(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.end).toBeCalled();
  });

  it('A contact edit request fails to be created', async () => {
    const body = {
      productOwner: {
        roleId: 1,
        id: 233,
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@example.com',
        githubId: 'jane1100',
      },
      technicalContact: {
        roleId: 2,
        id: 234,
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        githubId: 'john1100',
      },
    };
    const req = {
      params: { profileId: 118 },
      body,
    }

    client.query.mockResolvedValue({
      productOwner: {
        userId: 'jane1100',
        provider: 'github',
        email: 'jane.doe@example.com',
      },
      technicalContact: {
        userId: 'john1100',
        provider: 'github',
        email: 'john.doe@example.com',
      },
    });

    client.query.mockImplementation(() => { throw new Error() });

    await expect(requestProfileContactsEdit(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.end).not.toBeCalled();
  });
});
