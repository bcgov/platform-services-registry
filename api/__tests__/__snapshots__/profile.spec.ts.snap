// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile event handlers A list of quota options for a given profile is returned 1`] = `200`;

exports[`Profile event handlers A list of quota options for a given profile is returned 2`] = `
Array [
  "medium",
]
`;

exports[`Profile event handlers A profiles contacts are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      ",
      "values": Array [
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers A profiles contacts are returned 2`] = `200`;

exports[`Profile event handlers A profiles contacts are returned 3`] = `
Array [
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.436Z",
    "email": "jane@example.com",
    "firstName": "Jane",
    "githubId": "jane1100",
    "id": 1,
    "lastName": "Doe",
    "roleId": 1,
    "updatedAt": "2020-09-10T18:14:13.436Z",
  },
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.436Z",
    "email": "john@example.com",
    "firstName": "John",
    "githubId": "john1100",
    "id": 2,
    "lastName": "Doe",
    "roleId": 2,
    "updatedAt": "2020-09-10T18:14:13.436Z",
  },
]
`;

exports[`Profile event handlers A profiles quota size is returned 1`] = `200`;

exports[`Profile event handlers A profiles quota size is returned 2`] = `"small"`;

exports[`Profile event handlers Fetch a list of quota options for a given profile should throw 1`] = `"Unable to fetch allowed quota-sizes for profile 4"`;

exports[`Profile event handlers Fetch profiles quota size should throw 1`] = `"Unable fetch profile quota size with profile ID 4"`;

exports[`Profile event handlers Fetch single profile contacts should throw 1`] = `"Unable fetch profile contacts with profile ID 1"`;

exports[`Profile event handlers Fetch single profile contacts should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      ",
      "values": Array [
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Link a contact to profile should throw 1`] = `"Unable to add contact to profile"`;

exports[`Profile event handlers Link a contact to profile should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO profile_contact
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        4,
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Link a contact to profile successfully 1`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO profile_contact
          (profile_id, contact_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        4,
        1,
      ],
    },
  ],
]
`;

exports[`Profile event handlers Link a contact to profile successfully 2`] = `201`;

exports[`Profile event handlers Link a contact to profile successfully 3`] = `undefined`;

exports[`Profile event handlers Profile edit requests are returned 1`] = `200`;

exports[`Profile event handlers Profile edit requests are returned 2`] = `
Array [
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.304Z",
    "editObject": Object {
      "quota": "medium",
      "quotas": Object {
        "cpu": Object {
          "limits": "16",
          "requests": "8",
        },
        "memory": Object {
          "limits": "64Gi",
          "requests": "32Gi",
        },
        "storage": Object {
          "backup": "50Gi",
          "block": "100Gi",
          "capacity": "100Gi",
          "file": "100Gi",
        },
      },
    },
    "editType": "quotaSize",
    "id": 1,
    "natsContext": Object {
      "action": "edit",
      "displayName": "Health Gateway",
      "namespaces": Array [
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "elf077-tools",
          "namespaceId": 1,
        },
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "elf077-test",
          "namespaceId": 2,
        },
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "elf077-dev",
          "namespaceId": 3,
        },
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "elf077-prod",
          "namespaceId": 4,
        },
      ],
      "newDisplayName": "NULL",
      "productOwner": Object {
        "email": "jane@example.com",
        "provider": "github",
        "userId": "jane1100",
      },
      "profileId": 4,
      "quota": "small",
      "quotas": Object {
        "cpu": Object {
          "limits": "8",
          "requests": "4",
        },
        "memory": Object {
          "limits": "32Gi",
          "requests": "16Gi",
        },
        "storage": Object {
          "backup": "25Gi",
          "block": "50Gi",
          "capacity": "50Gi",
          "file": "50Gi",
        },
      },
      "technicalContact": Object {
        "email": "john@example.com",
        "provider": "github",
        "userId": "john1100",
      },
      "type": "standard",
    },
    "natsSubject": "registry_project_provisioning",
    "profileId": 1,
    "updatedAt": "2020-09-10T18:14:13.304Z",
  },
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.304Z",
    "editObject": Array [
      Object {
        "email": "jane@example.com",
        "firstName": "Jane",
        "githubId": "jane1100test",
        "id": 1,
        "lastName": "Doe",
        "roleId": 1,
      },
      Object {
        "email": "jim@example.com",
        "firstName": "John",
        "githubId": "john1100test",
        "id": 2,
        "lastName": "Doe",
        "roleId": 2,
      },
    ],
    "editType": "contacts",
    "id": 2,
    "natsContext": Object {
      "action": "edit",
      "displayName": "Project X",
      "namespaces": Array [
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "4ea35c-tools",
          "namespaceId": 13,
        },
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "4ea35c-test",
          "namespaceId": 14,
        },
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "4ea35c-dev",
          "namespaceId": 15,
        },
        Object {
          "clusters": Array [
            Object {
              "clusterId": 1,
              "name": "silver",
              "provisioned": true,
            },
          ],
          "name": "4ea35c-prod",
          "namespaceId": 16,
        },
      ],
      "newDisplayName": "NULL",
      "productOwner": Object {
        "email": "jane@example.com",
        "provider": "github",
        "userId": "jane1100test",
      },
      "profileId": 4,
      "quota": "small",
      "quotas": Object {
        "cpu": Object {
          "limits": "8",
          "requests": "4",
        },
        "memory": Object {
          "limits": "32Gi",
          "requests": "16Gi",
        },
        "storage": Object {
          "backup": "25Gi",
          "block": "50Gi",
          "capacity": "50Gi",
          "file": "50Gi",
        },
      },
      "technicalContact": Object {
        "email": "john@example.com",
        "provider": "github",
        "userId": "john1100test",
      },
      "type": "standard",
    },
    "natsSubject": "registry_project_provisioning",
    "profileId": 4,
    "updatedAt": "2020-09-10T18:14:13.304Z",
  },
]
`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 1`] = `202`;

exports[`Profile event handlers Request profile contacts edit with provisioner-related changes 2`] = `undefined`;

exports[`Profile event handlers Request profile quota size edit should throw due to db transaction issue 1`] = `"Unable to update quota-size for profile 4"`;

exports[`Profile event handlers Request profile quota size edit should throw due to invalid quota upgrade 1`] = `"Unable to update quota-size for profile 4"`;

exports[`Profile event handlers Request profile quota size edit successfully 1`] = `204`;

exports[`Profile event handlers Request profile quota size edit successfully 2`] = `undefined`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 1`] = `204`;

exports[`Profile event handlers Update profile contacts with non provisioner-related changes 2`] = `undefined`;
